### ISSUE TITLE
Bug: @typescript-eslint/no-unnecessary-condition false positive hiding actual issue

### ISSUE THREAD

**Issue by ryan-tote** [typescript-eslint/typescript-eslint#11316]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.8.2&fileType=.tsx&code=MYewdgzgLgBFCm0DyAjAVgLhgJXqATgCYA80%2BAlmAOYA0MYArgLYrz4DaAugHwwC8MAN4BfAFChIsANbwAnvxgBGAAwBuUaITJ07GbM4wA-IYFd1mxFFRpdczgDoADgwgALABQqAlKqA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokADThseIqVlUa9ZIxasm42rCZNEZFMknRCY8UwAmlfJTdLIKtUhtAF8QYKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false

### Repro Code

```TypeScript
const testObj: Record<string, number[]> = {}
const key = 10;

testObj[key] ??= [];

// the below code works fine for some reason which is why you also need 
// the rule @typescript-eslint/prefer-nullish-coalescing to be active for this bug to show
// if (!testObj[key]) {
//   testObj[key] = [];
// }

testObj[key].push(10);
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/prefer-nullish-coalescing": "error",
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "noUncheckedIndexedAccess": true
  }
}
```

### Expected Result

Shows an error showing that using a number for the key instead of a string is the issue.

### Actual Result

Shows the following error 

`Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined`

despite that not actually being the issue

**Comment by bradzacher** [comment]
This is working as expected - 

https://typescript-eslint.io/rules/no-unnecessary-condition/#possibly-undefined-indexed-access

**Comment by ryan-tote** [comment]
This says that "To get pessimistic, but correct, types for these cases, you can use TypeScript's [noUncheckedIndexedAccess compiler option](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess)" which has been enabled as can be seen in the above config. 

The above code works for keys that are strings but not for numbers. Are you sure this is expected behaviour @bradzacher ?

**Comment by ryan-tote** [comment]
For a code example

```typescript
const test: Record<string, number[]> = {};
// errors Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined.
test[12] ??= [];
// no error
test["12"] ??= [];
```

**Comment by bradzacher** [comment]
Sorry - I pattern matched the snippet and didn't notice the compiler options.

This is working as expected due to a limitation in TS. I'm on mobile so can't easily search for the relevant issue - but the TLDR is that TS reports the incorrect type for the left hand side of the nullish coalescing operator here - it reports the final type (I.e. The non-nullish type) so the rule incorrectly reports.

**Comment by kirkwaiblinger** [comment]
I think https://github.com/typescript-eslint/typescript-eslint/issues/6632 -> https://github.com/microsoft/TypeScript/issues/53303 are the existing issues around `??=` types

**Comment by ryan-tote** [comment]
Thanks both, this all makes sense now. Think I'm just going to avoid the `??=` syntax for these cases
